### PR TITLE
Added FactorizePhys

### PR DESCRIPTION
**Pull request by jnj256**
- Add folder - FactorizePhys/neural_methods/model/FactorizePhys **[Completed]**
- Add FactorizePhys/neural_methods/trainer/FactorizePhysTrainer.py **[Completed]**
- Update config.py file **[Completed]**
- Add training and inference config files **[Completed]**
- Add selected pretrained models. **[Completed]**
- [Optional] Add YOLO5Face detector that the repo uses (FactorizePhys/dataset/data_loader/face_detector) **[Completed]**
- [Optional] Add FactorizePhys/tools/output_signal_viz/batch_comparison.py for output signal plots generation for multiple models as a batch operation.**[Pending]**

### CODE REVIEW

**Review by yahskapar — CHANGES_REQUESTED** [review]
Thanks for the PR!

Few things that are missing  / seemed problematic during my testing:

* You need to add support for picking FactorizePhys as a model in `main.py`, within the `train_and_test()` and `test_only()`. functions.
* Also add to `neural_methods/trainer/__init__.py` a lien for `FactorizePhysTrainer`

I would also go ahead and remove any mentions of Y5F in the config files - I think we can replace RetinaFace with Y5F in a near future PR. I unfortunately haven't been able to test beyond that due to some CUDA-related issues, let me debug that a bit more later (may be unrelated to this PR, possibly specific to my machine).

**Comment by yahskapar** [comment]
If you have time and resources, I also recommend testing changes locally as a part of the toolbox, perhaps once with an inference config and once with a train config. It would be nice to see if things are breaking somehow across different machines...

File: neural_methods/model/FactorizePhys/test_FactorizePhys.py
[the remark this answers is not in the document]
Comment (jnj256) [reply neural_methods/model/FactorizePhys/test_FactorizePhys.py:1]: > Outside of running these files themselves after changing the config paths, is there any other planned way to use these test files (e.g., through configs)? No worries if not, can still leave these files in this PR as I think they can be useful.

In the FactorizePhys repo, all model files were augmented with a code-snippet to assess latency for each model. Though I agree with your suggestion that we can consider having such general test_files for each model separately and then to be able to execute them through config files. In this PR, it is not supported in this manner. It is still possible to run them directly as below:

``` bash
python -m neural_methods.model.FactorizePhys.test_FactorizePhys
```

**Comment by yahskapar** [comment]
> > If you have time and resources, I also recommend testing changes locally as a part of the toolbox, perhaps once with an inference config and once with a train config. It would be nice to see if things are breaking somehow across different machines...
> 
> This is pending for now. Once completed from my end, I'll update here.

Thanks, also one more thing @jnj256: for each instance of the device configuration in the config file, make sure that the default is set to `cuda:0` instead of `cuda:1`. Currently it's set to `cuda:1`, which will cause an error for most people who run this toolbox with a single GPU. `cuda:0` is the default in the other configs as well.

**Comment by yahskapar** [comment]
Also, I ran this on my machine and got the below result with `UBFC-rPPG_PURE_FactorizePhys_FSAM_Res.yaml`:

```
FFT MAE (FFT Label): 2.7931342690677967 +/- 1.3790434373891178
FFT RMSE (FFT Label): 10.954701564862207 +/- 8.175978575633126
FFT MAPE (FFT Label): 5.319840579003846 +/- 2.737428640110342
FFT Pearson (FFT Label): 0.8969699816717884 +/- 0.058556443302813746
FFT SNR (FFT Label): 3.5600835892137503 +/- 0.38396366175070534 (dB)
```

This doesn't seem to match the corresponding row of Table 2 in the paper [here](url), where the MAE is `0.48`. Am I potentially messing something up with the default config provided? Curious if you get a similar result with RetinaFace...maybe the usage of YOLOV5 is very important with this model.

**Comment by yahskapar** [comment]
> > > If you have time and resources, I also recommend testing changes locally as a part of the toolbox, perhaps once with an inference config and once with a train config. It would be nice to see if things are breaking somehow across different machines...
> > 
> > 
> > This is pending for now. Once completed from my end, I'll update here.
> 
> Thanks, also one more thing @jnj256: for each instance of the device configuration in the config file, make sure that the default is set to `cuda:0` instead of `cuda:1`. Currently it's set to `cuda:1`, which will cause an error for most people who run this toolbox with a single GPU. `cuda:0` is the default in the other configs as well.

Additionally, aside from doing this in the config file, also set `EXP_NAME` to be empty for each split configuration. It's fine if users change it to something else later, but setting something other than an empty string as a default effectively overrides the default naming that includes the pre-processing parameters (e.g., chunk length, face detection back-end, etc).

**Comment by yahskapar** [comment]
Definitely seems like Y5F makes a huge difference, below is the much closer result I got with [your repo](https://github.com/PhysiologicAILab/FactorizePhys):

```
FFT MAE (FFT Label): 0.521385063559322 +/- 0.17291570599033101
FFT RMSE (FFT Label): 1.4268612496710038 +/- 0.3487470223898108
FFT MAPE (FFT Label): 0.825317604282858 +/- 0.2956666697517955
FFT Pearson (FFT Label): 0.9982891950059382 +/- 0.007744475262706065
FFT SNR (FFT Label): 13.65674385665829 +/- 0.8034746791391728 (dB)
```

I propose we integrate it along with this PR, if possible - no worries if that means it takes more time to get this PR merged. I think it may be confusing to toolbox users if they try to reproduce your results without Y5F.

**Comment by yahskapar** [comment]
> > > > If you have time and resources, I also recommend testing changes locally as a part of the toolbox, perhaps once with an inference config and once with a train config. It would be nice to see if things are breaking somehow across different machines...
> > > 
> > > 
> > > This is pending for now. Once completed from my end, I'll update here.
> > 
> > 
> > Thanks, also one more thing @jnj256: for each instance of the device configuration in the config file, make sure that the default is set to `cuda:0` instead of `cuda:1`. Currently it's set to `cuda:1`, which will cause an error for most people who run this toolbox with a single GPU. `cuda:0` is the default in the other configs as well.
> 
> Additionally, aside from doing this in the config file, also set `EXP_NAME` to be empty for each split configuration. It's fine if users change it to something else later, but setting something other than an empty string as a default effectively overrides the default naming that includes the pre-processing parameters (e.g., chunk length, face detection back-end, etc).

Yet another minor thing to fix that I noticed, if you don't mind: make the MACC metric formatted like the other metrics when it's printed out (e.g., with `FFT` or `PEAK`) in [metrics.py](https://github.com/ubicomplab/rPPG-Toolbox/blob/main/evaluation/metrics.py). Also, unrelated, but fix this minor typo that I likely made on [this line](https://github.com/ubicomplab/rPPG-Toolbox/blob/main/evaluation/metrics.py#L193) (it should be PEAK SNR and Peak Label there).

Additionally, I recommend re-ordering where `MACC` appears as a metrics option in all config files. Currently, it's:

`METRICS: ['MAE', 'RMSE', 'MAPE', 'Pearson', 'SNR', 'BA', 'MACC']`

Which causes the `MACC` metric to be printed out after some lines related to the Bland-Altman plots. I propose doing instead doing:

`METRICS: ['MAE', 'RMSE', 'MAPE', 'MACC', 'Pearson', 'SNR', 'BA']`

Which should look a bit cleaner.

**Comment by yahskapar** [comment]
(I realize these are quite a few requested changes, so let me know if it would help to just merge this PR as is and do separate PRs, @jnj256. I should have some time to work on the changes I suggested sometime this weekend.)

**Comment by jnj256** [comment]
> (I realize these are quite a few requested changes, so let me know if it would help to just merge this PR as is and do separate PRs, @jnj256. I should have some time to work on the changes I suggested sometime this weekend.)

Thanks @yahskapar for your detailed analysis and testing of this PR. In my most recent commit, I've addressed all above converns that you raised. Please try to test all the changes and let me know if you encounter any issues.

**Review by yahskapar — APPROVED** [review]
LGTM, I have some minor clean-us in mind + would like to remove RetinaFace, but this can be done in a separate PR later. Thanks for your work with integrating FactorizePhys here, @jnj256, and great work again with the paper!

**Comment by jnj256** [comment]
> LGTM, I have some minor clean-us in mind + would like to remove RetinaFace, but this can be done in a separate PR later. Thanks for your work with integrating FactorizePhys here, @jnj256, and great work again with the paper!

Thanks to you @yahskapar for fully supporting and testing this integration.
